File: src/java_solutions/Solution00973.java
package java_solutions;

import java.util.*;

public class Solution00973 {
    public int[][] kClosest(int[][] points, int k) {
        // keep track of k minimum distances using max heap (replacing the max value if necessary)
        var maxHeap = new PriorityQueue<Point>(k, (a, b) -> b.distance - a.distance);

        // Calculate the distance from the origin for every point
        for (int[] point : points) {
            // No need to use sqrt() as comparing squared distances is sufficient and more efficient
            int distance = (int) (Math.pow(point[0], 2) + Math.pow(point[1], 2));
            if (maxHeap.size() >= k) {
                // add the point to max heap replacing the maximum value if necessary
                if (distance < maxHeap.peek().distance) {
                    maxHeap.poll();
                    maxHeap.offer(new Point(point, distance));
                }
            } else {
                maxHeap.offer(new Point(point, distance));
            }
        }

        var result = new int[k][2];
        // return the contents of the max heap
        int counter = 0;
        for (Point point : maxHeap) {
            result[counter++] = point.coords;
        }

        return result;
    }

    public static void main(String[] args) {
        var solution = new Solution00973();

    }
}

class Point {
    int[] coords;
    int distance;

    public Point(int[] coords, int distance) {
        this.coords = coords;
        this.distance = distance;
    }
}

File: src/Solution00973.kt
import java.util.PriorityQueue
import kotlin.math.pow
import kotlin.math.sqrt

data class Point(val distance: Double, val x: Int, val y: Int)

/**
 * Optimal solution: O(n log k)
 */
fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
    val maxHeap = PriorityQueue<Point> { a, b ->
        when {
            a.distance < b.distance -> 1
            a.distance > b.distance -> -1
            else -> 0
        }
    }
    // val maxHeap = PriorityQueue<Point> { a, b -> b.distance.compareTo(a.distance) }

    points.forEach {
        val newPoint = Point(getDistance(it), it[0], it[1])
        if (maxHeap.size < k) {
            maxHeap.offer(newPoint)
        } else if (maxHeap.peek().distance > newPoint.distance) {
            maxHeap.poll()
            maxHeap.offer(newPoint)
        }
    }

    val result = mutableListOf<IntArray>()
    while (maxHeap.isNotEmpty()) {
        val point = maxHeap.poll()
        result.add(intArrayOf(point.x, point.y))
    }
    return result.toTypedArray()
}

/**
 * Second attempt.
 */
fun kClosest2(points: Array<IntArray>, k: Int): Array<IntArray> {
    val minHeap = PriorityQueue<Point> { a, b ->
        when {
            a.distance < b.distance -> -1
            a.distance > b.distance -> 1
            else -> 0
        }
    }

    points.forEach {
        val newPoint = Point(getDistance(it), it[0], it[1])
        minHeap.offer(newPoint)
    }

    val result = mutableListOf<IntArray>()
    repeat(k) {
        val point = minHeap.poll()
        result.add(intArrayOf(point.x, point.y))
    }
    return result.toTypedArray()
}

/**
 * My first attempt.
 *
 * Positive aspects:
 * 1.The overall approach using a min-heap (PriorityQueue) is correct
 * and efficient.
 * 2.The distance calculation function is implemented correctly.
 * 3.The code handles the input and output formats properly.
 *
 * Areas for improvement:
 * 1.Using a map to store distances and points can lead to issues if
 * multiple points have the same distance. This could result in
 * losing some points.
 * 2.The current implementation has a time complexity of O(n log n)
 * due to adding all points to the heap. It can be optimized to O(n log k)
 * 3.Unnecessary use of Double for distances when Float would suffice.
 */
fun kClosest1(points: Array<IntArray>, k: Int): Array<IntArray> {
    val minHeap = PriorityQueue<Double>()
    val map = mutableMapOf<Double, IntArray>()

    points.forEach {
        val distance = getDistance(it)
        minHeap.add(distance)
        map[distance] = it
    }

    val result = mutableListOf<IntArray>()
    repeat(k) {
        map[minHeap.poll()]?.let {
            result.add(it)
        }
    }
    return result.toTypedArray()
}

private fun getDistance(point: IntArray): Double {
    return sqrt(point[0].toDouble().pow(2.0) + point[1].toDouble().pow(2.0))
}

fun main() {
    var points: Array<IntArray> = arrayOf(
        intArrayOf(1, 3),
        intArrayOf(-2, 2)
    )
    println(kClosest(points, 1).contentDeepToString()) // [[-2, 2]]

    points = arrayOf(
        intArrayOf(3, 3),
        intArrayOf(5, -1),
        intArrayOf(-2, 4)
    )
    println(kClosest(points, 2).contentDeepToString())  // [[3, 3], [-2, 4]]

    points = arrayOf(
        intArrayOf(0, 1),
        intArrayOf(1, 0)
    )
    println(kClosest(points, 2).contentDeepToString()) // [[0,1],[1,0]]
}
